lib/core/src/persist/sync.rs

```diff
@@ -259,4 +275,325 @@ impl Persister {
 
         Ok(())
     }
+
+    fn check_commit_update(con: &Connection, record_id: &str, last_updated_at: u32) -> Result<()> {
+        let mut stmt = Self::get_sync_outgoing_details_stmt(con)?;
+        let mut rows = stmt.query(named_params! {
+            ":id": record_id,
+        })?;
+
+        if let Some(row) = rows.next()? {
+            let sync_outgoing_details = Self::sql_row_to_sync_outgoing_details(row)?;
+
+            if sync_outgoing_details.updated_at > last_updated_at {
+                return Err(anyhow::anyhow!("Record has been updated while pulling"));
+            }
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn commit_receive_swap(
+        &self,
+        data: &ReceiveSyncData,
+        sync_state: SyncState,
+        last_updated_at: Option<u32>,
+    ) -> Result<()> {
+        let mut con = self.get_connection()?;
+        let tx = con.transaction_with_behavior(TransactionBehavior::Immediate)?;
+
+        if let Some(last_updated_at) = last_updated_at {
+            Self::check_commit_update(&tx, &sync_state.record_id, last_updated_at)?;
+        }
+
+        let params = named_params! {
+            ":id": &data.swap_id,
+            ":invoice": &data.invoice,
+            ":preimage": &data.preimage,
+            ":create_response_json": &data.create_response_json,
+            ":claim_fees_sat": &data.claim_fees_sat,
+            ":claim_private_key": &data.claim_private_key,
+            ":payer_amount_sat": &data.payer_amount_sat,
+            ":receiver_amount_sat": &data.receiver_amount_sat,
+            ":mrh_address": &data.mrh_address,
+            ":mrh_script_pubkey": &data.mrh_script_pubkey,
+            ":created_at": &data.created_at,
+            ":payment_hash": &data.payment_hash,
+            ":description": &data.description,
+            ":state": PaymentState::Created,
+        };
+        match self.fetch_receive_swap_by_id(&data.swap_id)?.is_some() {
```

Comment:
Executing here a query in a different connection inside a transaction will cause deadlock.